### PR TITLE
AP_Baro: rewritten BMP280 driver to fix glitches

### PR DESCRIPTION
**Pull request by night-ghost**
and allow to use non-standard BMP280 chip letter Y, found on some Matek boards

### CODE REVIEW

**Comment by night-ghost** [comment]
This is definitely sensor error, because it occures even on SPI bus. Seems that internal shadowing mechanism don't works, that why mode of sensor changed from normal to forced, when measuring intervai is directed by driver and no shadowing is needed. Moreover temperature glitches occures in the different sample than pressure glitches, so this is definitely not a bus error.

This driver was hard tested on MatekF405-CTR board which also has BMP280 on I2C, and it works just fine.

Current BMP280 driver was not tested at all because there was no supported boards with such sensor. There are long baro related discussions  on RCG and [here](http://www.ykoctpa.ru/groups/eye-in-a-sky/forum/topic/ardupilot-na-platax-openpilot-revolution-revomini/?topic_page=1&num=15) (in russian)

**Comment by sh83** [comment]
You have missed part, that updates pressure_filter:
_pressure_filter.apply(press);

File: libraries/AP_Baro/AP_Baro_BMP280.cpp [hunk 1]
```diff
@@ -126,24 +145,37 @@ bool AP_Baro_BMP280::_init()
...
 
 // transfer data to the frontend
 void AP_Baro_BMP280::update(void)
 {
     if (_sem->take_nonblocking()) {
-        if (!_has_sample) {
-            _sem->give();
-            return;
+        if (_has_sample) {
+            float pressure = _pressure_filter.getf();
```
Comment (sh83) [inline libraries/AP_Baro/AP_Baro_BMP280.cpp]: _pressure_filter is not updated. So driver didnt work at all at current from!

File: libraries/AP_Baro/AP_Baro_BMP280.h [hunk 2]
```diff
@@ -34,6 +34,12 @@ class AP_Baro_BMP280 : public AP_Baro_Backend
     int32_t _t_fine;
     float _pressure;
     float _temperature;
+    float _mean_temperature;
+    uint8_t _mask;
+    bool _letterY;
```
Comment (sh83) [inline libraries/AP_Baro/AP_Baro_BMP280.h]: Some comments are welcome. It is totally unclear that this means.

File: libraries/AP_Baro/AP_Baro_BMP280.cpp [hunk 3]
```diff
@@ -126,24 +145,37 @@ bool AP_Baro_BMP280::_init()
 void AP_Baro_BMP280::_timer(void)
 {
     uint8_t buf[6];
+    uint8_t sts;
+
+    // check status first
+    if(!_dev->read_registers(BMP280_REG_STATUS, &sts, sizeof(sts))) return;
+    if (sts & ((1<<3) | (1<<0)) ) { // if conversion is on or NVM copy in progress
```
[the remark this answers is not in the document]
Comment (night-ghost) [reply libraries/AP_Baro/AP_Baro_BMP280.cpp]: This requires a lots of additional defines, and don't changes how it works

File: libraries/AP_Baro/AP_Baro_BMP280.cpp (hunk 1 above)
Comment (night-ghost) [reply libraries/AP_Baro/AP_Baro_BMP280.cpp]: Oops... it was lost preparing PR to upstream. Fixed now

File: libraries/AP_Baro/AP_Baro_BMP280.h (hunk 2 above)
Comment (night-ghost) [reply libraries/AP_Baro/AP_Baro_BMP280.h]: names of variables are self explaining

File: libraries/AP_Baro/AP_Baro_BMP280.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sh83) [reply libraries/AP_Baro/AP_Baro_BMP280.cpp]: It needs just two defines, but makes code much more readable.

File: libraries/AP_Baro/AP_Baro_BMP280.cpp (hunk 1 above)
Comment (sh83) [reply libraries/AP_Baro/AP_Baro_BMP280.cpp]: Thank you

File: libraries/AP_Baro/AP_Baro_BMP280.h (hunk 2 above)
Comment (night-ghost) [reply libraries/AP_Baro/AP_Baro_BMP280.h]: see PR's description

**Comment by tridge** [comment]
> This is definitely sensor error, because it occures even on SPI bus. Seems that internal shadowing mechanism don't works, that why mode of sensor changed from normal to forced, when measuring intervai is directed by driver and no shadowing is needed. Moreover temperature glitches occures in the different sample than pressure glitches, so this is definitely not a bus error.

Can you give a log or console output showing the error? How often do they happen?
I haven't seen any glitches yet with my KakuteF4 which has a BMP280 on I2C.

**Comment by mirkix** [comment]
@night-ghost current driver is used by BeagleBone Blue (https://youtu.be/v2xWtlyYrtE )

**Comment by night-ghost** [comment]
>is there any documentation from Bosch on the differences with revision Y sensors?

I don't found. Also I asked the representative of Matek, but he does not have it either. But with new driver termocompensation of pressure works fine on Y revision so I think that formulas are correct.

>Can you give a log or console output showing the error? How often do they happen?

I don't know, but even one bad value cause bad behavior
https://www.rcgroups.com/forums/showthread.php?2784349-Ardupilot-on-OpenPilot-Revolution/page166#post39436619

You can read the full story from [here](https://www.rcgroups.com/forums/showthread.php?2784349-Ardupilot-on-OpenPilot-Revolution/page135#post39101438)

**Comment by tridge** [comment]
@night-ghost the log 1.BIN in that thread is bizarre. It shows a single 3250 meter baro glitch, but NKF1.PD is affected for a long time. That should not happen. The whole point of using a EKF is to get rid of spurious sensor data that is outside of the innovation gate.
I tested this in SITL and wasn't able to reproduce. I'd like to look at the exact version of ArduPilot that this happened for. The log is for ArduPlane V3.9.0-dev (597e0c04 2018-03-19), but git revision 597e0c04 doesn't exist anywhere I can find it. Can you tell me where I can find this exact git revision so I can try to find the problem?

**Comment by night-ghost** [comment]
> a single 3250 meter baro glitch, but NKF1.PD is affected for a long time

That's why I start to filter baro and compass data before EKF, because I seen the same behaviour sometimes - single glitch causes large EKF madness. 

>Can you tell me where I can find this exact git revision so I can try to find the problem?

Sorry but nowhere :( - rebasing kills old commits and creates new on its place, so there is no way to get a sources from that time now. But I think that this can be simulated by small piece of code which should alter baro data eg. each 30 seconds.

**Comment by tridge** [comment]
I tried simulating it in SITL with a 3250m glitch every 5 seconds, and the EKF correctly rejected it.
If you find some way to simulate this and reproduce the issue then let me know. We shouldn't just work around this with filters, we should fix this if its a real bug in the EKF code

**Comment by night-ghost** [comment]
BTW,  [This post](https://www.rcgroups.com/forums/showpost.php?p=39150082&postcount=2068) has a console log of real baro glitches. At the beginning 99294.38 is a correct value, but the filter is poisoned and need a time. At the end of log one can see real glitches - "Baro pressure error: mean 99304.93 got 28027.59". And yes, I don't know why the wrong value is so stable.

**Comment by night-ghost** [comment]
>We shouldn't just work around this with filters, we should fix this if its a real bug in the EKF code

EKF2 definitely has some issues in a recent code, many people [reported](http://www.ykoctpa.ru/groups/eye-in-a-sky/forum/topic/ardupilot-na-platax-openpilot-revolution-revomini/?topic_page=106&num=15#post-15561) (in russian) that copter with EKF2 flies glitchy, but I seen a PR about removing EKF2 so just adviceing to switch to EKF3 which works fine.

**Comment by tridge** [comment]
@night-ghost videos aren't really useful for diagnosing issues - much more useful is a dataflash log
I've received a MatekF405-CTR now, and I'm going to do some local testing to see if I can reproduce the issue from this PR

**Comment by night-ghost** [comment]
>videos aren't really useful for diagnosing issues

Yes I know, but this was in old times when I hadn't logs at all, so video just shows that issue exists.

>I've received a MatekF405-CTR now, and I'm going to do some local testing to see if I can reproduce the issue from this PR

we ordered 2 such boards, one comes with good baro letter K and one (my) with baro letter Y.

**Comment by night-ghost** [comment]
according to datasheet, letter K in 1st place of 2nd line means that this is ordinary production sensor. 

I don't desolder this board so don't know what is on the back side

**Comment by tridge** [comment]
@night-ghost I'm still reluctant to merge this change. If we get a single pressure value over 300000 in a BMP280 that isn't of type 'Y' then all readings would be wrong from then on. A single bus glitch would cause the aircraft to crash.
btw, are you absolutely sure that this is a BMP280? Or is it a different sensor with the same ID?

**Comment by night-ghost** [comment]
>If we get a single pressure value over 300000 in a BMP280 that isn't of type 'Y'

you are right, this should be checked after a filter and to set a time limit, will do. But I don't know more reliable way to distinguish between sensors :(

>btw, are you absolutely sure that this is a BMP280?

I absolutely don't sure, I just spoke with Matek representative who said "we bought a bunch of such sensors in Taobao".

**Comment by sh83** [comment]
Well this PR combines 3 different changes.
1. "Support" for Y chip revision
2. Use pressure and temp filter
3. Change measuring mode from NORMAL to FORCED.

The first change is the most problematic one. Actually there is some sensor with yp marking. See https://www.reddit.com/r/AskElectronics/comments/6r4h4j/bmp280_absolute_barometric_pressure_sensor_from/ But nobody knows what to do with it. Some samples work as normal bmp280, some samples dont even connect, some had different chip id, some requires correction. Most probably it was custom made for apple with different specs.
2. Second change is obvious, it could help sometimes in case of bus errors(bmp280 opposed to ms5811) dont have crc.
3. Third change is the intresting one. The reason behind this is the following paragraph from ds:
"3.10  Data  register shadowing In  normal  mode,  measurement  timing  is  not  necessarily  synchronized  to  readout.  This  means that  new  measurement  results  may  become  available  while  the  user  is  reading  the  results  from the  previous  measurement.  In  this  case,  shadowing  is  performed  in  order  to  guarantee  data consistency.  Shadowing  will  only  work  if  all  data  registers  are  read  in  a  single  burst  read. Therefore,  the  user  must  use  burst  reads  if  he  does  not  synchronize  data  readout  with  the measurement  cycle.  Using  several  independent  read  commands  may  result  in inconsistent  data"
So we cannot garantee data consistency in case of normal mode and stop conditions on i2c bus. It is better to use FORCED mode. It is the same approach inav uses in it bmp280 driver.

**Comment by night-ghost** [comment]
I removed letter Y logick from this PR to not affect normal sensors in any way, so let Matek use binaries from my fork

**Comment by night-ghost** [comment]
>So we cannot garantee data consistency in case of normal mode and stop conditions on i2c bus

But things are much worse, and we regularily got errors even without stop, even on SPI - see link to full story above. This errors can't take place in 'forced' mode

**Comment by peterbarker** [comment]
Is anyone still interested in chasing this PR?

It is currently conflicting.  And old :-)

**Pull request closed without merging**